src/main/java/com/mongodb/jdbc/MongoSQLResultSetMetaData.java

```diff
@@ -2,57 +2,144 @@
 
 import java.sql.ResultSetMetaData;
 import java.sql.SQLException;
-import java.sql.SQLFeatureNotSupportedException;
-import org.bson.BsonType;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
 
 public class MongoSQLResultSetMetaData extends MongoResultSetMetaData implements ResultSetMetaData {
+    private static class NameSpace {
+        String datasource;
+        String columnLabel;
 
-    public MongoSQLResultSetMetaData(MongoResultDoc metadataDoc) {
-        super(metadataDoc);
+        NameSpace(String datasource, String columnLabel) {
+            this.datasource = datasource;
+            this.columnLabel = columnLabel;
+        }
+    }
+
+    private static class DatasourceAndIndex {
+        String datasource;
+        int index;
+
+        DatasourceAndIndex(String datasource, int index) {
+            this.datasource = datasource;
+            this.index = index;
+        }
+    }
+
+    // A mapping from columnLabel name to datasource name and index.
+    private Map<String, DatasourceAndIndex> columnLabels;
+    // A mapping from index position to NameSpace (datasource, columnLabel).
+    private List<NameSpace> columnIndices;
+    // A mapping from index position to ColumnTypeInfo.
+    private List<MongoSQLColumnTypeInfo> columnTypeInfo;
+
+    // This gets the datasource for a given columnLabel, and is used
+    // in MongoSQLResultSet to retrieve data by label.
+    String getDatasource(String columnLabel) {
+        return columnLabels.get(columnLabel).datasource;
+    }
+
+    private static void assertObjectSchema(MongoJsonSchema schema) throws SQLException {
+        if (schema.bsonType == null
+                || !schema.bsonType.equals("object")
+                || schema.properties == null) {
+            throw new SQLException("ResultSetMetaData json schema must be object with properties");
+        }
+    }
+
+    private void processDataSource(MongoJsonSchema schema, String datasource) throws SQLException {
+        MongoJsonSchema datasourceSchema = schema.properties.get(datasource);
+        assertObjectSchema(datasourceSchema);
+
+        Object[] columns = datasourceSchema.properties.keySet().toArray();
```

Comment:
```suggestion
        String[] columns = datasourceSchema.properties.keySet().toArray(new String[0]);
```
We could get the columns as Strings from the start and avoid the cast in the loop.